rust/datafusion/src/datasource/datasource.rs

```diff
@@ -48,9 +66,19 @@ pub trait TableProvider {
         &self,
         projection: &Option<Vec<usize>>,
         batch_size: usize,
+        filters: &[Expr],
     ) -> Result<Arc<dyn ExecutionPlan>>;
 
     /// Returns the table Statistics
     /// Statistics should be optional because not all data sources can provide statistics.
     fn statistics(&self) -> Statistics;
+
+    /// Tests whether the table provider can make use of a filter expression
+    /// to optimise data retrieval.
+    fn test_filter_pushdown(
```

Comment:
Maybe `supports_filter_pushdown`?

Reply:
I feel this makes us evaluate the expression twice: once for testing filter support and once at scan time. Moreover, it is likely that different partitions will have different `TableProviderFilterPushDown` values. It would be great optimize out filters on that finer granularity 🚀 But that would imply doing that optimization in the physical plan.

Reply:
> Maybe `supports_filter_pushdown`?

That's definitely better, other people seem to like it as well, sold :p

> I feel this makes us evaluate the expression twice: once for testing filter support and once at scan time. Moreover, it is likely that different partitions will have different `TableProviderFilterPushDown` values. It would be great optimize out filters on that finer granularity 🚀 But that would imply doing that optimization in the physical plan.

Yeah, I think this depends on whether it's preferable to:
- offer extensibility for providers, allowing them to interpret filters however they like
- promote some notion of external data chunk into an actual DataFusion concept (I need to read through your catalogue proposal doc properly!) and get providers to return a list of available chunks with their bounds, then having DataFusion perform the high-level filtering where possible, and pass that more selective list of chunk IDs back for actual data retrieval

And personally I see pros and cons for both of those options (flexibility is cool, but sometimes dangerous).

I'm not familiar enough with partitioning in DataFusion yet to quite understand what you mean for the different pushdown support return values across partitions; I would expect that the filter pushdown status is usually a property of the column in question? e.g. depending on whether a column tracks sufficient statistics and the provider storage mechanism allows for it

Reply:
> I'm not familiar enough with partitioning in DataFusion yet to quite understand what you mean for the different pushdown support return values across partitions

Sorry, I was not very clear! Let me try to do better 😄 

When the logical plan is converted, DataFusion allows the target physical plan to specify a number of partitions. This typically allows you to use that many threads (or workers) to execute your physical plan. In the context of datasources, this is often the number of files, as they can always be decoded separately and in parallel. But it could be any other relevant partitioning of your data. My point, is that you will often have the statistics at the granularity of the file <=> DataFusion partitions. 

So the workflow here will be the following:
- In the logical plan, you are evaluating the pushed-down expressions on the entire datasource (all of its files) to know if some of the expressions can be simplified out (using `supports_filter_pushdown()`). To achieve this in your TableProvider implem, you will need to merge all the statistics of your files into one that characterizes the entire datasource, then select the expressions.
- You will likely need to do the same expression pruning again at the partition level (so in the physical expression). Indeed each file will have more focused (= accurate) statistics that you will want to take advantage of, because you will likely be able to prune out more parts of the filter expression.

My question is whether it is worth adding the complexity of trying to simplify the filtering expression here in the logical plan when it will better done at the partition level in the physical plan.

Reply:
@rdettai  -- I think it is reasonable to assume something (I am purposely being vague here) could be done at logical planning time that could save execution time later on. While I don't fully follow the scenario you are describing, I can see how in some cases you won't save much by trying to do partition pruning at the logical planning level.

However, I can think of scenarios where you might be able to (like deciding what hosts to send a plan to, for example). You might be right that that should be done at Physical Planning time. 

It is my opinion that DataFusion is moving sufficiently fast at this point that we are discovering the requirements as we go that having a potentially more complicated API is an acceptable tradeoff, especially given how well this one is documented.